### ISSUE TITLE
[SPI, dv] Missing SPI pass through test with two stages read pipeline

### ISSUE THREAD

**Issue by sha-ron** [lowRISC/opentitan#24597]
### Description

SPI top level test (not unit level) that can run on VCS simulation and configures CMD_INFO.read_pipeline_mode to 0x2.
https://opentitan.org/book/hw/ip/spi_device/doc/registers.html#cmd_info--read_pipeline_mode

**Comment by moidx** [comment]
@vogelpi is it possible to assess if we have any targeted test cases for this mode of operation?

**Comment by moidx** [comment]
This is currently impacting `earlgrey_1.0.0`. We would like to be able to get coverage here to be able to provide feedback to Nuvoton as part of the characterization exercise. 

This will also be needed for gen2.

**Comment by vogelpi** [comment]
@a-will and @engdoreis provided some very useful feedback on this. There are two relevant tests for this feature.

SiVal
---

The `//sw/device/tests:spi_passthru_test` **SiVal** test **exercises all pipeline modes**. This can be seen from the following files:
- The device side C code [`spi_passthru_test.c`](https://github.com/lowRISC/opentitan/blob/48f0fb8747cc7b0a078f90d7e6bad8aa5258be01/sw/device/tests/spi_passthru_test.c#L30) uses a volatile constant `kReadPipelineMode` set to 0 (this corresponds to 0 stages in [cmd_info--read_pipeline_mode](https://opentitan.org/book/hw/ip/spi_device/doc/registers.html#cmd_info--read_pipeline_mode).
- The host side code Rust code [`spi_passthru.rs`](https://github.com/lowRISC/opentitan/blob/fb62e4880551fb4bea4c02aff86b3a1d145f5987/sw/host/tests/chip/spi_device/src/spi_passthru.rs#L499) overwrites this constant and cycles through all the different settings.

This test has been passing on FPGA and A1/A2 silicon for quite some time now according to the SiVal spreadsheet.
 
DV
---

For the `chip_sw_spi_device_pass_through` **DV** test, the situation differs between `master` and `earlgrey_1.0.0`. On both branches, the pipeline mode setting is a random configuration knob in the DV agents for SPI, see e.g. https://github.com/lowRISC/opentitan/blob/d095b772b6a65e8318924fe326d12a874eb210fb/hw/dv/sv/spi_agent/spi_flash_cmd_info.sv#L15 and https://github.com/lowRISC/opentitan/blob/d095b772b6a65e8318924fe326d12a874eb210fb/hw/dv/sv/spi_agent/spi_item.sv#L22

On **`master`**, the chip-level sequence [chip_sw_spi_passthrough_vseq.sv](https://github.com/lowRISC/opentitan/blob/master/hw/top_earlgrey/dv/env/seq_lib/chip_sw_spi_passthrough_vseq.sv) uses the spi_item with the randomized read_pipeline config knob and overwrites the constant similar to the host Rust code above. The sequence contains logic:
  - to detect when the pipeline mode is enabled https://github.com/lowRISC/opentitan/blob/f74d4ab331c5ec9dd53c821bedf8a71304f24375/hw/top_earlgrey/dv/env/seq_lib/chip_sw_spi_passthrough_vseq.sv#L130-L140 (Note that this change got merged after the creation of this issue, it was part of this PR https://github.com/lowRISC/opentitan/pull/24733)
  - to reduce the SPI clock for the gate level simulations due to certain settings for the read pipeline mode (the comment even references this very issue here) https://github.com/lowRISC/opentitan/blob/f74d4ab331c5ec9dd53c821bedf8a71304f24375/hw/top_earlgrey/dv/env/seq_lib/chip_sw_spi_passthrough_vseq.sv#L17-L21 as well as
  - to enforce a value of 2 for the read pipeline mode in GLS
https://github.com/lowRISC/opentitan/blob/f74d4ab331c5ec9dd53c821bedf8a71304f24375/hw/top_earlgrey/dv/env/seq_lib/chip_sw_spi_passthrough_vseq.sv#L159-L163

On **`ealrgrey_1.0.0`, the the chip-level sequence [chip_sw_spi_passthrough_vseq.sv](https://github.com/lowRISC/opentitan/blob/earlgrey_1.0.0/hw/top_earlgrey/dv/env/seq_lib/chip_sw_spi_passthrough_vseq.sv) does NOT overwrite the read_pipeline_mode on the C side ([spi_passthrough_test.c](https://github.com/lowRISC/opentitan/blob/earlgrey_1.0.0/sw/device/tests/sim_dv/spi_passthrough_test.c) does not contain a corresponding constant). The clock is still reduced for gate level simulations mentioning 30 MHz would be too fast for values different from 0 or 1 of read_pipeline_mode https://github.com/lowRISC/opentitan/blob/5a96d2d802eceb0224fa76b8c7e7ada6762eb60e/hw/top_earlgrey/dv/env/seq_lib/chip_sw_spi_passthrough_vseq.sv#L17-L21

But I can't see the test configuring 2 for the read_pipeline_mode. The test seems always run with value 0 for read_pipeline_mode. @sha-ron can you please confirm this?

---

So to summarize:
- It seems we have test coverage for this feature in SiVal and tests have been passing on FPGA and A1/A2 silicon.
- We have test coverage chip-level simulations on `master` but not on `earlgrey_1.0.0`. 

What we should do in my view:
1. Cherry-pick PR https://github.com/lowRISC/opentitan/pull/24733 which added support for this feature in the chip-level simulation over to `earlgrey_1.0.0` and run this on on the `ealgrey_1.0.0` branch.
2. Clarify which setting was actually tested in GLS for A1/A2 and what the results were for the different frequencies. @sha-ron , could you please help us with that?
3. Resolve the various TODOs in `chip_sw_spi_passthrough_vseq.sv` once we have clarified the situation.

Does this sound reasonable?

**Comment by sha-ron** [comment]
GLS simulation was done on earlgrey_1.0.0 branch and used read_pipline_mode=0  with 24MHz.

**Comment by vogelpi** [comment]
Thanks @sha-ron , and I assume the test passed with these settings right?

**Comment by sha-ron** [comment]
> Thanks [@sha-ron](https://github.com/sha-ron) , and I assume the test passed with these settings right?

yes:)

**Comment by sha-ron** [comment]
Hi @vogelpi, do you if the silicon was checked with read_pipline_mode=2 and at which frequency?

**Comment by vogelpi** [comment]
> Hi [@vogelpi](https://github.com/vogelpi), do you if the silicon was checked with read_pipline_mode=2 and at which frequency?

Based on https://github.com/lowRISC/opentitan/blob/48f0fb8747cc7b0a078f90d7e6bad8aa5258be01/sw/device/tests/spi_passthru_test.c#L278-L291

The Sival test on Silicon runs at 24 MHz (96 MHz / 4). It tests all values for read_pipeline_mode (including value 2). The host side of the test (the rust code linked above) overwrites the configuration defined in the device code.

**Comment by vogelpi** [comment]
Going back to my list in https://github.com/lowRISC/opentitan/issues/24597#issuecomment-3744042017:

- [x] Cherry-pick PR https://github.com/lowRISC/opentitan/pull/24733 which added support for this feature in the chip-level simulation over to `earlgrey_1.0.0` and run this on on the `ealgrey_1.0.0` branch. -> This has been done now, thanks @rswarbrick (see #29368 )
- [x] Clarify which setting was actually tested in GLS for A1/A2 and what the results were for the different frequencies. -> @sha-ron confirmed that the setting `read_pipline_mode`=0 was tested in GLS with a clock frequency of 24MHz
- [x] Resolve the various TODOs in `chip_sw_spi_passthrough_vseq.sv` (on [earlgrey_1.0.0](https://github.com/lowRISC/opentitan/blob/earlgrey_1.0.0/hw/top_earlgrey/dv/env/seq_lib/chip_sw_spi_passthrough_vseq.sv) and on [master](https://github.com/lowRISC/opentitan/blob/master/hw/top_earlgrey/dv/env/seq_lib/chip_sw_spi_passthrough_vseq.sv)) once we have clarified the situation. See below. -> Done in #30129

There are two TODOs:
   - Line 161: In GLS, the `read_pipeline_mode` setting is forced to 2 motivated by GLS failing for frequencies above 24 MHz otherwise.
   - Line 18: in GLS, the frequency setting of 30 MHz is excluded motivated by `read_pipeline_mode` settings of 0 and 1 would not work with 30 MHz.
    
  So only 24 MHz and 6 MHz can be run in GLS at the moment and only `read_pipeline_mode` = 2 is tested. However, we know the following:

  - From @sha-ron : 24 MHz with `read_pipeline_mode` = 0 works in GLS.
  - From SiVal: 24 MHz with all `read_pipeline_mode` settings on actual silicon.

   
  What we should do IMO to resolve these TODOs:
  - [x] Remove the foreach loop on Line 204 which cycles through all clock settings. -> Done in #30129
  - [x] Remove the `ifndef GATE_LEVEL` on Line 17, i.e., allow testing 6, 24 and 30 MHz also in GLS. Randomly pick one of these three settings per test execution. -> Done in #30129
  - [x] Change the constraint for the `read_pipeline_mode` setting (Line 32) to enforce a value of two for the 30 MHz clock setting but randomly choose any of the values 0, 1, 2 for other clock settings. -> Done in #30129
  - [x] Remove the `ifdef GATE_LEVEL` on Line 159 including the enforcing of a specific `read_pipeline_mode` setting.
    - See https://github.com/lowRISC/opentitan/pull/29557
  - [x] Print the combination of clock and read_pipeline_mode test in DV as well as have the core print the read_pipeline_mode setting to convince ourselves we test the rights thing. -> Done in #30129

Does this sound reasonable @sha-ron  and @rswarbrick ? Would you be able to rerun a couple of gate level simulations to confirm the changes work as expected @sha-ron ?

**Comment by sha-ron** [comment]
It sounds reasonable to me. It will be a lot of effort to run gate level simulations now since it is not synched with the head of earlgrey_1.0.0 branch and there are also issues with the air-gapped Bazel.  But the changes can be done and it will be checked on the next gate level simulations.

**Comment by vogelpi** [comment]
Thanks for the feedback @sha-ron , do you have another functional branch at the moment, such as https://github.com/lowRISC/opentitan/tree/earlgrey_1.0.0_A2_presilicon? If the diff isn't too large, we could maybe provide a commit for you to try it out. I would prefer to resolve this for good and move forward afterwards.

**Comment by sha-ron** [comment]
Unfortunately we don't have another functional branch (and currently also no access to GLS). I think the issue can be closed for now. Worst case we will open something new if there will be a need in the future.

**Comment by rswarbrick** [comment]
I believe that this will be resolved by #30129.

**Comment by vogelpi** [comment]
I am closing this issue, @rswarbrick 's PR here #30129 got merged and backported to ealrgrey_1.0.0 here #30198 . I am closing this issue. Thanks again Rupert!
